Fix rendering on JupyterHub 4.1: respect `absolute_url` for protocol determination, support `data-absolute-url`

krassowski opened the pull request:
- [x] issues: fixes #14002
   - which together with https://github.com/holoviz/pyviz_comms/pull/125
   - helps to fix https://github.com/holoviz/panel/issues/7039
- [x] tests passed (see below for manual testing instructions)
- [ ] release document entry (if new feature or API change)

This PR:
-  fixes `ws`/`wss` determination when `absolute_url` is passed
- supports reading the absolute URL from `data-absolute-url` of an iframe as `srcdoc` iframes cannot pass it otherwise (e.g. there is no motion of URL parameters for these)

| Before | After |
|--|--|
| ![Screenshot from 2024-07-29 17-16-06](https://github.com/user-attachments/assets/c898db35-aacb-4601-b8e5-ac6b55db9be5) | ![Screenshot from 2024-07-29 17-29-59](https://github.com/user-attachments/assets/3a5d2057-e128-4aa0-95db-d16cef2b8a6b) |


<details>

<summary>How to test without installing JupyterHub</summary>

1. Install pyviz_comms with https://github.com/holoviz/pyviz_comms/pull/125 applied
2. Run `jupyter --paths` and choose a writable config path from the `config:` section, e.g. `/home/username/my-env/etc/jupyter`
3. Open (create if needed) a `/home/username/my-env/etc/jupyter/jupyter_server_config.py`, paste the following:
   ```py
    c.LabApp.tornado_settings = {
        'hub_prefix': '/hub',
        'hub_host': 'test',
        'user': 'test',
        'headers': {'Content-Security-Policy': "frame-ancestors 'none'"}
    }
   ```
4. Start `jupyter lab`
5. Open dev tools Network tab, click on a request and expand Response Headers, verify that CSP for frame ancestors is set to none:
   ![image](https://github.com/user-attachments/assets/f446bacb-5be2-4bef-9508-a4b844adb8bf)
6. Ensure that page data includes `hubPrefix` by viewing the source of JupyterLab page
   ![Screenshot from 2024-07-30 10-06-57](https://github.com/user-attachments/assets/cf74d7a2-a2e2-40bd-b876-7b294725047e)
7. Follow reproduction steps in https://github.com/holoviz/panel/issues/7039


</details>

krassowski committed:
Fix rendering on JupyterHub 4.1

mattpap commented on bokehjs/src/lib/embed/server.ts:
```diff
@@ -8,9 +8,10 @@ import type {EmbedTarget} from "./dom"
-    protocol = "wss:"
+  // if in an `srcdoc` iframe, try to get the absolute URL
+  // from the `data-absolute-url` attribute if not passed explicitly
+  if (absolute_url === undefined && frameElement !== null && (frameElement as HTMLIFrameElement).dataset.absoluteUrl !== undefined) {
+    absolute_url = (frameElement as HTMLIFrameElement).dataset.absoluteUrl
```
Let's remove all this casting with:
```suggestion
  if (absolute_url === undefined && frameElement instanceof HTMLElement && frameElement.dataset.absoluteUrl !== undefined) {
    absolute_url = frameElement.dataset.absoluteUrl
  }
```
or alternatively use `HTMLIFrameElement` instead of `HTMLElement`, though I'm not sure if we need to be that specific.

krassowski replied:
My thinking was the same here! But it does not work. This is because `frameElement` is an `HTMLElement`, but not `window.HTMLElement`. It is an `HTMLElement` from the context of the parent frame. I can make this work without type cast by checking for the right instance - which is just a bit more code - but `frameElement` by definition is an `HTMLElement` ("element" is in the name).

krassowski replied:
Do you still think I should modify it here or do you agree with the type cast here?

krassowski replied:
For example, using `frameElement instanceof window.parent.HTMLElement` can result in:

```
DOMException: Failed to read a named property 'HTMLElement' from 'Window': Blocked a frame with origin "http://localhost:5151" from accessing a cross-origin frame.
```

krassowski replied:
Resolved in https://github.com/bokeh/bokeh/pull/14003/commits/c639a85a7a571d0075e78b6804613d9af6ecea99

mattpap commented on bokehjs/src/lib/embed/server.ts:
```diff
@@ -21,6 +22,11 @@ export function _get_ws_url(app_path: string | undefined, absolute_url: string |
+  if (loc.protocol == "https:") {
+    protocol = "wss:"
+  }
+
```
Also let's get rid of a mutable variable:
```suggestion
  const protocol = loc.protocol == "https:" ? "wss:" : "ws:"
```

krassowski committed:
Get rid of the mutable variable

Co-authored-by: Mateusz Paprocki <mattpap@gmail.com>

mattpap commented:
We need to be able to test this somehow, preferably outside of jupyter. Perhaps an example showing `<iframe>` embedding would suffice (under `examples/server`).

krassowski committed:
Add an example

krassowski committed:
Do not use type casting, define a custom type guard

krassowski committed:
Reduce size of the negative example iframe

krassowski commented:
@mattpap thank you for your review (dziękuję!) - I've added an example which both demonstrates the problem and a solution under `examples/server/embed`:

![image](https://github.com/user-attachments/assets/836faf11-e9e3-45f6-ab81-3e9201c18032)

I also added a dedicated type guard to remove the explicit type cast (I do not like them either!).

krassowski committed:
Lint

bryevdv commented:
`examples/server` should only contain subdirectories to match the docs structure, not individual examples. Also "embed" by itself is too general, there are lots of concepts about embedding. Possible suggestions:

* `examples/server/app/iframe_embed`
* somewhere under `examples/integration`

krassowski committed:
Move to `app/iframe_embed`

krassowski commented:
Thank you @bryevdv, I moved it to `examples/server/app/iframe_embed` following your suggestion.

bryevdv commented:
@mattpap will have to comment on the test screenshot error 
```
ERROR: Path for Bokeh server application does not exist: /home/runner/work/bokeh/bokeh/bokeh_server.py
tests/test_examples.py::test_server_examples[/home/runner/work/bokeh/bokeh/examples/server/app/iframe_embed-example12-config12] 
[FAIL] Unhandled rejection at:  ProtocolError: Cannot take screenshot with 0 width.
    at /home/runner/work/bokeh/bokeh/bokehjs/node_modules/chrome-remote-interface/lib/chrome.js:94:35
    at Chrome._handleMessage (/home/runner/work/bokeh/bokeh/bokehjs/node_modules/chrome-remote-interface/lib/chrome.js:272:17)
    at WebSocket.<anonymous> (/home/runner/work/bokeh/bokeh/bokehjs/node_modules/chrome-remote-interface/lib/chrome.js:240:22)
    at WebSocket.emit (node:events:518:28)
    at Receiver.receiverOnMessage (/home/runner/work/bokeh/bokeh/bokehjs/node_modules/chrome-remote-interface/node_modules/ws/lib/websocket.js:1070:20)
    at Receiver.emit (node:events:518:28)
    at Receiver.dataMessage (/home/runner/work/bokeh/bokeh/bokehjs/node_modules/chrome-remote-interface/node_modules/ws/lib/receiver.js:517:14)
    at Receiver.getData (/home/runner/work/bokeh/bokeh/bokehjs/node_modules/chrome-remote-interface/node_modules/ws/lib/receiver.js:435:17)
    at Receiver.startLoop (/home/runner/work/bokeh/bokeh/bokehjs/node_modules/chrome-remote-interface/node_modules/ws/lib/receiver.js:143:22)
    at Receiver._write (/home/runner/work/bokeh/bokeh/bokehjs/node_modules/chrome-remote-interface/node_modules/ws/lib/receiver.js:78:10) {
  request: {
    method: 'Page.captureScreenshot',
    params: { format: 'png', clip: [Object] },
    sessionId: undefined
  },
  response: { code: -32000, message: 'Cannot take screenshot with 0 width.' }
}

FAILED
```

krassowski commented:
Looking at:

https://github.com/bokeh/bokeh/blob/54252f8ba722a01a249df0aa603899cc1bbb2349/tests/examples.yaml#L18-L31

It seems that similar examples (e.g. examples `server_document`, `json_item`) are skipped. Should this one be skipped too?

bryevdv commented:
> Should this one be skipped too?

Let's just go with that option for now

krassowski committed:
Skip `iframe_embed` for now

krassowski commented:
It did not help. I think this is because the top-level file is not an example of Bokeh server with a single handler, but a flask application which embeds code which connects to the server running in a subprocess.

I am not sure how to proceed, possibly moving it elsewhere would be better. If a maintainer would prefer to rehash it please do push to my branch.

krassowski commented:
Sorry, I was wrong it did help, another job failed which appears unrelated. I will restart CI by pushing an empty commit to see if it helps.

krassowski committed:
Retrigger CI

krassowski commented:
Previously `Bokeh-CI / unit-test (macos-latest, 3.12)` failed and now `Bokeh-CI / unit-test (macos-latest, 3.11)` failed - it looks like the macos tests are flaky, is this correct?

bryevdv commented:
Yes there is one server test that probably ought to get a retry policy or similar.

bryevdv commented:
@krassowski I really appreciate your patience, I have one more request. After thinking about it, I would prefer this live under `examples/server/api` instead of `app`. All the examples under `app` are meant to be run with `bokeh serve` by the user, this example is much more like the embedding examples under `api`.  I don't want to create confusion where someone tries to run `bokeh serve iframe_embed` (like all the other directory examples there) and it does not work.

krassowski committed:
Move to `examples/server/api`

krassowski commented:
@bryevdv make sense - done, and it looks all tests passed!

bryevdv commented:
Thanks @krassowski !

krassowski commented:
Amazing, thank you! Is there a chance for backporting this PR or will it be only available in 3.6?

bryevdv commented:
It seems like it would be easy to backport, so I have added the label. However, that just means that _if_  there is a new release on previous branches, then this is a candidate to go in it. It's reasonably likely there will end up being more earlier point releases, but I can't guarantee it.

Edit: though looking at other PRs marked for backport, it seems fairly like there will be at some point.

bryevdv commented:
@krassowski this fix is available now in 3.5.2
